src/guest_config/src/lib.rs

```diff
@@ -1,5 +1,233 @@
+use std::fs;
+use std::fs::File;
+use std::io::{BufWriter, Read};
+use std::path::Path;
+
+use cpuid::Cpuid;
+use logger::{debug, error, info, warn};
+
+use crate::cpu::cpu_config::{
+    CpuConfigurationAttribute, CustomCpuConfiguration, CustomCpuConfigurationApiRequest,
+};
+use crate::cpu::cpu_symbolic_engine::CPU_FEATURE_INDEX_MAP;
+
 /// Contains types used to configure guest vCPUs.
 pub mod cpu;
+
+const JSON_FILE_FORMAT_EXTENSION: &str = "json";
+
+/// Errors associated with processing CPU configuration
+#[derive(Debug, thiserror::Error)]
+pub enum GuestConfigurationError {
+    /// Error while configuring CPU features via CPUID
+    #[error("Failed to configure CPU features via CPUID")]
+    CpuId,
+    /// Error while configuration model-specific registers
+    #[error("Error while configuring CPU features via model-specific registers")]
+    MSR,
+    /// Invalid file path specified
+    #[error("Invalid file path specified - {0}")]
+    InvalidFilePath(String),
+    /// Invalid file type provided
+    #[error(
+        "Invalid file type specified. Only JSON or extensionless (binary) is accepted - [{0}]"
+    )]
+    InvalidFileType(String),
+    #[error("Invalid file format provided. Not able to deserialize guest configuration - [{0}]")]
+    JsonError(serde_json::Error),
+    /// Unknown/Undefined CPU feature name
+    #[error("Unknown or undefined CPU feature name")]
+    UndefinedCpuFeatureName,
+    #[error("CPU feature override for [{0}] is not supported")]
+    UnsupportedCpuFeatureOverride(String),
+    /// Opening or reading the file was unsuccessful
+    #[error("Unable to use file specified [{0}]. See error [{1}]")]
+    IOError(String, std::io::Error),
+    /// Binary file operation was unsuccessful
+    #[error("Error handling binary file [{0}]. See error [{1}]")]
+    BinaryFileOperationError(String, bincode::Error),
+}
+
+/// Takes a file referred to by a template path string and
+/// compiles the json types referred to in the template
+/// into binary objects, converting from a
+/// CustomCpuConfigurationApiRequest to CustomCpuConfiguration
+pub fn compile_template_file(
+    template_file_path_string: &str,
+) -> Result<CustomCpuConfiguration, GuestConfigurationError> {
+    compile_template_string(
+        fs::read_to_string(template_file_path_string).expect(
+            format!(
+                "Unable to read template file [{}]",
+                template_file_path_string
+            )
+            .as_str(),
+        ),
+    )
+}
+
+/// Take a JSON string of CustomCpuConfigurationApiRequest and
+/// converts to the JSON-deserialized type, CustomCpuConfiguration
+pub fn compile_template_string(
+    custom_cpu_config_request_string: String,
+) -> Result<CustomCpuConfiguration, GuestConfigurationError> {
+    match CustomCpuConfigurationApiRequest::try_from(custom_cpu_config_request_string) {
+        Ok(request) => return compile_template_request(&request),
+        Err(err) => Err(err),
+    }
```

Comment:
This can be simplified.
```suggestion
    let request = CustomCpuConfigurationApiRequest::try_from(custom_cpu_config_request_string)?;
    compile_template_request(&request)
```